File: Problems/Leetcode/001-1000/101-200/139.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        vector<int> ok(n + 1);
        ok[0] = 1;

        for (int i = 0; i < n; ++i) {
            for (string& t : wordDict) {
                if (ok[i] && s.substr(i, t.size()) == t) {
                    ok[i + t.size()] = 1;
                }
            }
        }

        // set<string> S(wordDict.begin(), wordDict.end());
        // for (int i = 1; i <= n; ++i) {
        //     for (int j = 0; j < i; ++j) {
        //         if (ok[j] && S.contains(s.substr(j, i - j))) {
        //             ok[i] = 1;
        //             break;
        //         }
        //     }
        // }

        return ok[n];
    }
};

int main() {
    Solution sol;

    return 0;
}

File: Problems/Leetcode/001-1000/101-200/139.py
from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        stringSet = set(wordDict)

        n = len(s)
        f = [False] * (n + 1)
        f[0] = True
        for i in range(1, n + 1):
            for j in range(i):
                if f[j] and s[j:i] in stringSet:
                    f[i] = True
                    break
        return f[n]
